sinabs/backend/dynapcnn/dynapcnn_network.py

```diff
@@ -1,99 +1,306 @@
         ----------
-            snn: sinabs.Network
-                SNN that determines the structure of the `DynapcnnNetwork`
-            input_shape: None or tuple of ints
-                Shape of the input, convention: (features, height, width)
-                If None, `snn` needs an InputLayer
-            dvs_input: bool
-                Does dynapcnn receive input from its DVS camera?
-            discretize: bool
-                If True, discretize the parameters and thresholds.
-                This is needed for uploading weights to dynapcnn. Set to False only for
-                testing purposes.
+        - snn (nn.Module): a  implementing a spiking network.
+        - input_shape (tuple): a description of the input dimensions as `(features, height, width)`.
+        - dvs_input (bool): wether or not dynapcnn receive input from its DVS camera.
+        - discretize (bool): If `True`, discretize the parameters and thresholds. This is needed for uploading 
+            weights to dynapcnn. Set to `False` only for testing purposes.
+        - weight_rescaling_fn (callable): a method that handles how the re-scaling factor for one or more `SumPool2d` projecting to
+            the same convolutional layer are combined/re-scaled before applying them.
+
+        Notes
+        ----------
+        Some of the properties defined within the class constructor are meant to be temporary data structures handling the conversion
+        of the `snn` (the original `nn.Module`) into a set of `DynapcnnLayer`s composing a `DynapcnnNetwork` instance. Once their role
+        in preprocessing `snn` is finished, all required data to train/deploy the `DynapcnnNetwork` instance is within `self._dcnnl_edges`
+        (the connectivity between each `DynapcnnLayer`/core), `self._layers_mapper` (every `DynapcnnLayer` in the network) and `self._merge_points`
+        (the `DynapcnnLayer`s that need a `Merge` input). Thus, the following private properties are delted as last step of the constructor:
+
+        - self._graph_tracer
+        - self._sinabs_edges
+        - self._sinabs_modules_map
+        - self._nodes_name_remap
+        - self._nodes_to_dcnnl_map
+        - self._dynapcnn_layers
+        - self._entry_nodes
         """
         super().__init__()
 
-        # This attribute stores the location/core-id of each of the DynapcnnLayers upon placement on chip
-        self.chip_layers_ordering = []
+        # TODO for now the graph part is not taking into consideration DVS inputs.
+        # check if dvs input is expected.
+        dvs_input = False
+        self.dvs_input = dvs_input
+        self.input_shape = input_shape
+
+        assert len(self.input_shape) == 3, "infer_input_shape did not return 3-tuple"
+
+        # computational graph from original PyTorch module.
+        self._graph_tracer = NIRtoDynapcnnNetworkGraph(
+            snn,
+            torch.randn((batch_size, *self.input_shape)))                # needs the batch dimension.
```

Comment:
Yes, it seems from the [documentation](https://github.com/neuromorphs/NIRTorch/blob/main/nirtorch/graph.py) that the batch dim. is not necessary so it can be dropped. The `self._get_nodes_io_shapes()` also ignores the batch dim. I think I kept it because the original `DynapcnnNetwork` implementation required it.

Reply:
Okay, I see now in the code for `extract_torch_graph` that it passes the sample data as-is to the module. Because sinabs requires 4D input, we won't get around having a batch/time dimension. In principle it shouldn't matter to us what that dimension is, but because the network is likely to contain `IAFSqueeze` layers, they will expect a specific value. Maybe we can try to infer this, in order to make the `batch_size` argument of `DynapcnnNetwork` optional.

Reply:
It should work without batch_size now. Please check @Willian-Girao